src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/CodeBackflower.cs

```diff
@@ -0,0 +1,244 @@
+using System.Threading.Tasks;
+using Microsoft.DotNet.Darc.Models.VirtualMonoRepo;
+using Microsoft.DotNet.DarcLib.Helpers;
+using Microsoft.Extensions.Logging;
+
+#nullable enable
+namespace Microsoft.DotNet.DarcLib.VirtualMonoRepo;
+
+public interface IVmrBackflower
+{
+    Task<List<VmrIngestionPatch>> CreateBackflowPatchesAsync(string mappingName, CancellationToken cancellationToken);
+
+    Task BackflowAsync(
+        BackflowAction action,
+        string repoName,
+        NativePath targetDirectory,
+        IReadOnlyCollection<AdditionalRemote> additionalRemotes,
+        CancellationToken cancellationToken);
+}
+
+public enum BackflowAction
+{
+    CreatePatches,
+    ApplyPatches,
+}
+
+/// <summary>
+/// This class is responsible for taking changes done to a repo in the VMR and backflowing them into the repo.
+/// It only makes patches/changes locally, no other effects are done.
+/// </summary>
+public class CodeBackflower : IVmrBackflower
+{
+    private readonly IVmrInfo _vmrInfo;
+    private readonly ISourceManifest _sourceManifest;
+    private readonly ILocalGitRepo _localGitClient;
+    private readonly IVmrPatchHandler _vmrPatchHandler;
+    private readonly IWorkBranchFactory _workBranchFactory;
+    private readonly IFileSystem _fileSystem;
+    private readonly ILogger<CodeBackflower> _logger;
+
+    public CodeBackflower(
+        IVmrInfo vmrInfo,
+        ISourceManifest sourceManifest,
+        ILocalGitRepo localGitClient,
+        IVmrPatchHandler vmrPatchHandler,
+        IWorkBranchFactory workBranchFactory,
+        IFileSystem fileSystem,
+        ILogger<CodeBackflower> logger)
+    {
+        _vmrInfo = vmrInfo;
+        _sourceManifest = sourceManifest;
+        _localGitClient = localGitClient;
+        _vmrPatchHandler = vmrPatchHandler;
+        _workBranchFactory = workBranchFactory;
+        _fileSystem = fileSystem;
+        _logger = logger;
+    }
+
+    public async Task<List<VmrIngestionPatch>> CreateBackflowPatchesAsync(string mappingName, CancellationToken cancellationToken)
+    {
+        (string vmrSourceSha, string vmrTargetSha, IVersionedSourceComponent repo) = await GetMappingInformation(mappingName);
+        return await CreateBackflowPatchesAsync(repo, vmrSourceSha, vmrTargetSha, cancellationToken);
+    }
+
+    public async Task BackflowAsync(
+        BackflowAction action,
+        string mappingName,
+        NativePath repoDirectory,
+        IReadOnlyCollection<AdditionalRemote> additionalRemotes,
+        CancellationToken cancellationToken)
+    {
+        (string vmrSourceSha, string vmrTargetSha, IVersionedSourceComponent repo) = await GetMappingInformation(mappingName);
+
+        var patches = await CreateBackflowPatchesAsync(repo, vmrSourceSha, vmrTargetSha, cancellationToken);
+        if (patches.Count == 0)
+        {
+            return;
+        }
+
+        if (action == BackflowAction.CreatePatches)
+        {
+            var message = new StringBuilder();
```

Comment:
Can we put this in a method and just call it `CreatePatches` and call it here?

Reply:
This bit does not create the patches though.
This bit is a special behaviour when we ONLY want to create patches and not apply them because in that case, we want to print the paths so that people can apply those themselves.

Reply:
I will add a comment though

Reply:
Ah true, we're doing that in a different method